### PR TITLE
Cleanup avatar related code

### PR DESCRIPTION
**Pull request by CarlSchwan**
- Move event listener to new event handling
- Add typing almost everywhere
- Fix inconsistent interface parameter

### CODE REVIEW

File: lib/private/Avatar/GuestAvatar.php [hunk 1]
```diff
@@ -53,17 +51,13 @@ public function __construct(string $userDisplayName, LoggerInterface $logger) {
 
 	/**
 	 * Tests if the user has an avatar.
-	 *
-	 * @return true Guests always have an avatar.
 	 */
-	public function exists() {
-		return true;
+	public function exists(): bool {
+		return true; // Guests always have an avatar.
```
Comment (come-nc) [inline lib/private/Avatar/GuestAvatar.php:56]:
```suggestion
		// Guests always have an avatar.
		return true;
```

File: lib/private/User/Listeners/UserDeletedListener.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+use OCP\EventDispatcher\Event;
+use OCP\EventDispatcher\IEventListener;
+use OCP\User\Events\UserDeletedEvent;
+use OCP\Files\NotFoundException;
+use OCP\IAvatarManager;
+use Psr\Log\LoggerInterface;
+
+/**
+ * @template-implements IEventListener<\OCP\User\Events\UserDeletedEvent>
```
Comment (come-nc) [inline lib/private/User/Listeners/UserDeletedListener.php:34]:
```suggestion
 * @template-implements IEventListener<UserDeletedEvent>
```

File: lib/private/User/Listeners/UserDeletedListener.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		if (!($event instanceof UserDeletedEvent)) {
+			return;
+		}
+		
+		$user = $event->getUser();
+
+		// Delete avatar on user deletion
+		try {
+			$avatar = $this->avatarManager->getAvatar($user->getUID());
+			$avatar->remove();
```
Comment (come-nc) [inline lib/private/User/Listeners/UserDeletedListener.php:55]:
```suggestion
			$avatar->remove(true);
```
Isn’t the $silent option made for this case? If not where is it used?

File: lib/private/Avatar/Avatar.php [hunk 4]
```diff
@@ -134,7 +124,7 @@ protected function getAvatarVector(int $size): string {
 	 * Generate png avatar from svg with Imagick
 	 *
 	 * @param int $size
-	 * @return string|boolean
+	 * @return string|false
```
Comment (blizzz) [inline lib/private/Avatar/Avatar.php:127]: maybe this can be changed to null? Then we'd win a return type.

File: lib/private/User/Listeners/UserDeletedListener.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		} catch (NotFoundException $e) {
+			// no avatar to remove
+		} catch (\Exception $e) {
+			// Ignore exceptions
+			$this->logger->info('Could not cleanup avatar of ' . $user->getUID());
```
Comment (blizzz) [inline lib/private/User/Listeners/UserDeletedListener.php:60]: perhaps add the exception to the log

File: tests/lib/Avatar/AvatarManagerTest.php [hunk 6]
```diff
@@ -101,7 +101,7 @@ public function testGetAvatarInvalidUser() {
 	}
 
 	public function testGetAvatarForSelf() {
-		$user = $this->createMock(IUser::class);
+		$user = $this->createMock(\OC\User\User::class);
```
Comment (blizzz) [inline tests/lib/Avatar/AvatarManagerTest.php:104]: could be imported

**Review by blizzz — COMMENTED** [review]
small suggestions

File: tests/lib/Avatar/AvatarManagerTest.php (hunk 6 above)
Comment (ChristophWurst) [reply tests/lib/Avatar/AvatarManagerTest.php:104]: was the switch from interface to implementation necessary?

Comment (ChristophWurst) [reply tests/lib/Avatar/AvatarManagerTest.php:104]: nvm, ``\OC\Avatar\UserAvatar::__construct`` needs the implementation.

**Comment by blizzz** [comment]
*hrumph*

> @since or @deprecated tag is needed in PHPDoc for OCP\Color::stepCalc

File: tests/lib/Avatar/AvatarManagerTest.php [hunk 7]
```diff
@@ -101,7 +102,7 @@ public function testGetAvatarInvalidUser() {
 	}
 
 	public function testGetAvatarForSelf() {
-		$user = $this->createMock(IUser::class);
+		$user = $this->createMock(User::class);
```
Comment (blizzz) [inline tests/lib/Avatar/AvatarManagerTest.php:105]: why was that necessary?

Comment (come-nc) [reply tests/lib/Avatar/AvatarManagerTest.php:105]: Because of https://github.com/nextcloud/server/pull/32973/files#diff-69c3b1c8e4b48cfe15a5f3aa2a72b84ed0688dd29bb9b60012fc2ee6400b8ab1R64 , UserAvatar constructor needs a User instance, before it was documented as phpdoc but now it’s an actual type check.

**Review by come-nc — APPROVED** [review]
